### PR TITLE
= str #22602 short path for Source.fromFuture and fromFutureSource an…

### PR DESCRIPTION
**Pull request by He-Pin**
refs https://github.com/akka/akka/issues/

### CODE REVIEW

**Comment by Hawstein** [comment]
Link to the correct issue. 

Refs #22602

**Comment by Hawstein** [comment]
You can add some tests for those completed future case.

**Comment by hepin1989** [comment]
@Hawstein I will let team to decide is these changes nice to have,then I will add test:)

**Comment by johanandren** [comment]
Looks good! Tests exercising the happy/failuer paths would be great though.

File: akka-stream-tests/src/test/scala/akka/stream/scaladsl/FutureFlattenSourceSpec.scala [hunk 1]
```diff
@@ -24,14 +24,42 @@ class FutureFlattenSourceSpec extends StreamSpec {
...
+        Source.fromFutureSource(Future.successful(underlying))
+          .toMat(Sink.seq)(Keep.both)
+          .run()
+
+      // should complete as soon as inner source has been materialized
+      sourceMatVal.futureValue should ===("foo")
+      sinkMatVal.futureValue should ===(List(1, 2, 3))
+    }
+
+    "emmit no elements before the future of source successful" in assertAllStagesStopped {
```
Comment (johanandren) [inline akka-stream-tests/src/test/scala/akka/stream/scaladsl/FutureFlattenSourceSpec.scala]: emit

**Review by johanandren — APPROVED** [review]
LGTM

File: akka-stream-tests/src/test/scala/akka/stream/scaladsl/FutureFlattenSourceSpec.scala (hunk 1 above)
Comment (hepin1989) [inline akka-stream-tests/src/test/scala/akka/stream/scaladsl/FutureFlattenSourceSpec.scala]: Will change,thanks

**Review by patriknw — APPROVED** [review]
LGTM

**Pull request merged**